File: php/problem23.php
<?php
//solution: 4179871
echo eulerProblem23() . "\n";

function eulerProblem23() {
    $arr = range(0,28123);
    for($i = 0; $i < count($arr); ++$i) {
        $is = isAbundant($i);
        if($is > 0) {
            $arr[$i] = $i;
        } else {
            $arr[$i] = 0;
        }
    }
    
    $sum = 0;
    for($i = 1; $i <= 20161; ++$i) {
        if(!isSumOfAbundantNumbers($arr, $i)) {
            echo $i . "\n";
            $sum += $i;
        }
    }
    return $sum;
}

function isSumOfAbundantNumbers($arr, $num) {
    if($num > 48 && $num % 2 == 0) {
        return 1;
    }
    
    for($i = 2; $i < count($arr); ++$i) {
        if($arr[$i] == 0) {
            continue;
        }
        $diff = $num - $arr[$i];
        if($diff < 0 || $diff > count($arr)) {
            return 0;
        }
        
        if($arr[$diff] != 0) {
            return 1;
        }
    }
    
    return 0;
}

function sumOfProperDivisors($num) {
    $sum = 1;
    for($i = 2; $i <= sqrt($num); ++$i) {
        if($num % $i == 0) {
            $sum += $i;
            if($num / $i != $i) {
                $sum += $num / $i;
            }
        }
    }
    return $sum;
}

function isAbundant($num) {
    $sum = sumOfProperDivisors($num);
    if($sum > $num) {
        return $sum;
    }
    
    return 0;
}

?>

File: python/problem23.py
#!/usr/bin/python
#solution: 4179871

import math

def proper_divisors(n):
    divs = [1]
    i = 2
    while i <= math.sqrt(n):
        if n % i == 0:
            divs.append(i)
            divs.append(n / i)
            if i == math.sqrt(n):
                divs.pop()
        i += 1
    return divs

def euler_problem_23():
    abun = {}
    for i in xrange(2, 28123):
        if i < sum(proper_divisors(i)): # abundant
            abun[i] = True
    total = 0
    i = 0
    while i < 28123:
        total += i
        for a in abun.keys():
            diff = i - a
            if diff <= 0:
                break
            if diff in abun:
                total -= i
                break
        i += 1
    return total 
    

print euler_problem_23()
